analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.PatternMatching.cs

```diff
@@ -287,5 +288,246 @@ public void DeclarationPattern_Var_Deconstruction_DoesNotSetConstraint()
+        [DataRow("nullableBoolNull", "true", null)]  // FN. Should be false.
+        [DataRow("nullableBoolNull", "false", null)] // FN. Should be false.
+        [DataRow("nullableBoolUnknown", "true", null)]
+        [DataRow("nullableBoolUnknown", "false", null)]
+        public void ConstantPatternSetBoolConstraint(string variableName, string isPattern, bool? expectedBoolConstraint)
+        {
+            const string variableDeclarations = @"
+var objectNotNull = new object();
+var objectNull = (object)null;
+var objectUnknown = Unknown<object>();
+var nullableBoolTrue = (bool?)true;
+var nullableBoolFalse = (bool?)false;
+var nullableBoolNull = (bool?)null;
+var nullableBoolUnknown = Unknown<bool?>();
+";
+            ValidateSetBoolConstraint(variableDeclarations, variableName, isPattern, OperationKindEx.ConstantPattern, expectedBoolConstraint);
+        }
+
+        [DataTestMethod]
+        [DataRow("objectNotNull", "{ }", true)]
+        [DataRow("objectNull", "{ }", false)]
+        [DataRow("objectUnknown", "{ }", null)]
+        [DataRow("stringNotNull", "{ }", true)]
+        [DataRow("stringNotNull", "{ Length: 0 }", null)]
+        [DataRow("stringNull", "{ Length: 0 }", false)]
+        [DataRow("stringNotNull", "{ Length: var length }", true)] // only deconstruction
+        public void RecursivePatternPropertySubPatternSetBoolConstraint(string variableName, string isPattern, bool? expectedBoolConstraint)
+        {
+            const string variableDeclarations = @"
+var objectNotNull = new object();
+var objectNull = (object)null;
+var objectUnknown = Unknown<object>();
+var stringNotNull = new string('c', 1);  // Make sure, we learn 's is not null'
+var stringNull = (string)null;
+";
+            ValidateSetBoolConstraint(variableDeclarations, variableName, isPattern, OperationKindEx.RecursivePattern, expectedBoolConstraint);
+        }
+
+#if NET
+
+        [DataTestMethod]
+        [DataRow("recordNotNull", "(A: 1, B: 2)", null)]
+        [DataRow("recordNotNull", "(A: var a, B: _)", true)]
+        [DataRow("recordNull", "(A: 1, B: 2)", false)]
+        [DataRow("recordNull", "(A: var a, B: _)", false)]
+        [DataRow("recordUnknown", "(A: var a, B: _)", null)]
+        public void RecursivePatternDeconstructionSubpatternSetBoolConstraint(string variableName, string isPattern, bool? expectedBoolConstraint)
+        {
+            const string variableDeclarations = @"
+var recordNotNull = new R(1, 2);
+var recordNull = (R)null;
+var recordUnknown = Unknown<R>();
+";
+            ValidateSetBoolConstraint(additionalTypes: "record R(int A, int B);", variableDeclarations, variableName, isPattern, OperationKindEx.RecursivePattern, expectedBoolConstraint);
+        }
+
+#endif
+
+        [DataTestMethod]
+        [DataRow("objectNull", "var a", true)]
+        [DataRow("objectNotNull", "var a", true)]
+        [DataRow("objectUnknown", "var a", null)] // FN. Should be "true". Some patterns always match.
```

Comment:
I would delete this one. While it could be `True`, the `null` is also correct. We didn't learn anything, we didn't take any decision => moving on in all directions. It should be fine.
```suggestion
        [DataRow("objectUnknown", "var a", null)]
```

Reply:
We would miss the unreachable part here:
```cs
if (some.Really.Long.Expression is var x)
{
  // always executed
} else
{
  // Unreachable
}